crates/bevy_ecs/src/bundle.rs

```diff
@@ -14,120 +14,185 @@ use bevy_ecs_macros::all_tuples;
+///
+/// For this reason, `bevy` does not store which bundles were used to create a given entity.
+/// There is also intentionally no [`Query`] filter which detects whether an entity contains the components
+/// of a bundle.
+/// To select multiple components in a query, you should specify the components exactly.
+///
+/// ## Removal
+///
+/// Bundles can also be used to remove components from an entity.
+///
+/// Removing a bundle from an entity will remove the components of the set it
+/// represents from the entity.
+/// A value of the [`Bundle`] type may then be collected containing these component values,
+/// and returned.
+///
+/// # Implementors
+///
+/// A [`Component`] can be added to or removed from a world.
+/// Therefore, every type which implements [`Component`] also implements [`Bundle`].
+///
+/// Additionally, [Tuples](`tuple`) of bundles are also [`Bundle`] (with up to 15 bundles).
+/// These bundles contain the items of the 'inner' bundles.
+/// This is a convenient shorthand which is primarily used when spawning entities.
+/// For example, spawning an entity using the bundle `(SpriteBundle {...}, PlayerMarker)`
+/// will spawn an entity with components required for a 2d sprite, and the `PlayerMarker` component.
+///
+/// [`unit`], otherwise known as [`()`](`unit`), is a [`Bundle`] containing no components (since it
+/// can also be considered as the empty tuple).
+/// This can be useful for spawning large numbers of empty entities using
+/// [`World::spawn_batch`](crate::world::World::spawn_batch)
+///
+/// Tuple bundles can be nested, which can be used to create an anonymous bundle with more than
+/// 15 items.
+/// However, in most cases where this is required, there is a [`derive@Bundle`] derive instead.
+/// This is used to create your own bundles from a struct of other [`Bundle`]s.
+/// As explained above, this includes any [`Component`] type, and other derived bundles:
 ///
-/// You can nest bundles using the `#[bundle]` attribute:
 /// ```
 /// # use bevy_ecs::{component::Component, bundle::Bundle};
 ///
 /// #[derive(Component)]
-/// struct X(i32);
-/// #[derive(Component)]
-/// struct Y(u64);
+/// struct XPosition(i32);
 /// #[derive(Component)]
-/// struct Z(String);
+/// struct YPosition(i32);
 ///
 /// #[derive(Bundle)]
-/// struct A {
-///     x: X,
-///     y: Y,
+/// struct PositionBundle {
+///     // A bundle can contain components
+///     x: XPosition,
+///     y: YPosition,
 /// }
 ///
 /// #[derive(Bundle)]
-/// struct B {
-///     #[bundle]
-///     a: A,
-///     z: Z,
+/// struct NamedPointBundle {
+///     // Or other bundles
+///     a: PositionBundle,
+///     // In addition to more components
+///     z: PointName,
 /// }
+///
+/// #[derive(Component)]
+/// struct PointName(String);
 /// ```
 ///
 /// # Safety
 ///
+/// Note: Manual implementations of this trait are *not* permitted. That is, there may be arbitrary additional
```

Comment:
Just curious: how is it possible to forbid manual trait implementations?

Reply:
Precisely by having the safety comments be incomplete (hence the 'that is').

Since they cannot know that they satisfy all the arbitrary additional safety conditions, they can't prove that their implementation is safe.
Therefore, they don't meet the conditions to write their `unsafe impl Bundle for MyType`.

Reply:
Oh, I thought this was a hard prohibition at the language level. I think then that it should be worded as “must not”. It is much more common, and “prohibited” may be interpreted as “not possible”.

Reply:
Alternatively, you can apply the [C-SEALED] API guideline, preventing downstream implementation.

[C-SEALED]: https://rust-lang.github.io/api-guidelines/future-proofing.html#sealed-traits-protect-against-downstream-implementations-c-sealed

Reply:
I don't follow?
How would sealing this trait work whilst still allowing the derive macro?

Reply:
Oh, you're right, the macro would still produce an implementation and wouldn't work...

Reply:
Yep; we can't seal this while allowing the derive macro to work.